pkg/groupsync/groups.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package groupsync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+)
+
+// ReadGroupClient provides read operations for a group system.
+type ReadGroupClient interface {
+	// Descendants retrieve all users (children, recursively) of a group.
+	Descendants(ctx context.Context, groupID string) ([]*User, error)
+
+	// GetGroup retrieves the Group with the given ID.
+	GetGroup(ctx context.Context, groupID string) (*Group, error)
+
+	// GetMembers retrieves the direct members (children) of the group with given ID.
+	GetMembers(ctx context.Context, groupID string) ([]Member, error)
+
+	// GetUser retrieves the User with the given ID.
+	GetUser(ctx context.Context, userID string) (*User, error)
+}
+
+// WriteGroupClient provides write operations for a group system.
+type WriteGroupClient interface {
+	// SetMembers replaces the members of the group with the given ID with the given members.
+	SetMembers(ctx context.Context, groupID string, members []Member) error
+}
+
+// ReadWriteGroupClient provides both read and write operations for a group system.
+type ReadWriteGroupClient interface {
+	ReadGroupClient
+	WriteGroupClient
+}
+
+// OneToManyGroupMapper maps group IDs to lists of group IDs.
+type OneToManyGroupMapper interface {
+	// AllGroupIDs returns the set of groupIDs being mapped (the key set).
+	AllGroupIDs(ctx context.Context) ([]string, error)
+
+	// ContainsGroupID returns whether this mapper contains a mapping for the given group ID.
+	ContainsGroupID(ctx context.Context, groupID string) (bool, error)
+
+	// MappedGroupIDs returns the list of group IDs mapped to the given group ID.
+	MappedGroupIDs(ctx context.Context, groupID string) ([]string, error)
+}
+
+// UserMapper maps a user ID to another user ID.
+type UserMapper interface {
+	// MappedUserID returns the user ID mapped to the given user ID.
+	MappedUserID(ctx context.Context, userID string) (string, error)
+}
+
+// User represents a user in a group system.
+type User struct {
+	// ID is the user's ID in the group system.
+	ID string `json:"id,omitempty"`
+	// Attributes represent arbitrary attributes about the user
+	// in the given group system. This field is typically set by
+	// the corresponding ReadGroupClient when retrieving the user.
+	Attributes any `json:"attributes,omitempty"`
```

Comment:
Should this be `map[string]any`?

Reply:
I considered this, but then I felt like leaving it as `any` gives the most flexibility for the implementing `GroupReader` to choose what makes the most sense for that Group System.

For example, maybe the implementing `GroupReader` wraps an existing client library (e.g. GitHub) and that client returns structs for its Groups and Users. If we leave the type as `any` then the `GroupReader` can simply set the `Attributes` field to that struct. If we require `map[string]any` then it has to convert the struct to a `map[string]any`. Not the biggest deal but seemed like an unnecessary restriction.

I'm not too familiar with the details of JSON marshalling/unmarshing in GO though. If there is a compelling reason to use `map[string]any` here I'm open to it.

Reply:
Using `any` is fine. In practice, `map[string]any` is often more commonly used for unknown JSON objects because:

- Most JSON data tends to be objects at the top level.
- It provides a balance between flexibility and structure.
- It's easier to iterate over and access key-value pairs.
- You still maintain flexibility for nested values (which are of type any).